src/Features/LanguageServer/Protocol/Api/InProcLanguageServer.cs

```diff
@@ -0,0 +1,404 @@
+
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Api
+{
+    internal class InProcLanguageServer : IAsyncDisposable
+    {
+        private readonly ServerCapabilities _serverCapabilities;
+
+        protected readonly JsonRpc JsonRpc;
+        protected readonly RequestDispatcher RequestDispatcher;
+        protected readonly RequestExecutionQueue Queue;
+        protected readonly ILspWorkspaceRegistrationService WorkspaceRegistrationService;
+        protected readonly IAsynchronousOperationListener Listener;
+        protected readonly ILspLogger Logger;
+        protected readonly string? ClientName;
+
+        /// <summary>
+        /// Server name used when sending error messages to the client.
+        /// </summary>
+        private readonly string _userVisibleServerName;
+
+        /// <summary>
+        /// Server name used when capturing error telemetry.
+        /// </summary>
+        protected readonly string TelemetryServerName;
+
+        // Set on first LSP initialize request.
+        protected ClientCapabilities? _clientCapabilities;
+
+        // Fields used during shutdown.
+        private bool _shuttingDown;
+        private Task? _errorShutdownTask;
+
+        internal bool HasShutdownStarted => _shuttingDown;
+
+        internal InProcLanguageServer(
+            AbstractRequestDispatcherFactory requestDispatcherFactory,
+            JsonRpc jsonRpc,
+            ServerCapabilities serverCapabilities,
+            ILspWorkspaceRegistrationService workspaceRegistrationService,
+            IAsynchronousOperationListenerProvider listenerProvider,
+            ILspLogger logger,
+            string? clientName,
+            string userVisibleServerName,
+            string telemetryServerTypeName)
+        {
+            RequestDispatcher = requestDispatcherFactory.CreateRequestDispatcher();
+
+            _serverCapabilities = serverCapabilities;
+            WorkspaceRegistrationService = workspaceRegistrationService;
+            Logger = logger;
+
+            JsonRpc = jsonRpc;
+            JsonRpc.AddLocalRpcTarget(this);
+
+            Listener = listenerProvider.GetListener(FeatureAttribute.LanguageServer);
+            ClientName = clientName;
+
+            _userVisibleServerName = userVisibleServerName;
+            TelemetryServerName = telemetryServerTypeName;
+
+            Queue = new RequestExecutionQueue(logger, workspaceRegistrationService, userVisibleServerName, TelemetryServerName);
+            Queue.RequestServerShutdown += RequestExecutionQueue_Errored;
+        }
+
+        /// <summary>
+        /// Handle the LSP initialize request by storing the client capabilities and responding with the server
+        /// capabilities.  The specification assures that the initialize request is sent only once.
+        /// </summary>
+        [JsonRpcMethod(Methods.InitializeName, UseSingleObjectParameterDeserialization = true)]
```

Comment:
Thinking ahead a bit, if we made the factory interface public, would this type be made public? Would these members stay public, or are they still an implementation detail? Does StreamJsonRpc let these still be internal?

Reply:
I would ideally not make these public.  There is a streamjsonrpc AllowNonPublicMethods option - https://github.com/microsoft/vs-streamjsonrpc/blob/main/doc/recvrequest.md#invocation-of-non-public-methods

But I don't know if it is problematic that it allows any non-public method?

I'm not sure if the type actually needs to be public - I can see instances where the host may want to intentionally dispose of the server?  But maybe doing it through streamjsonrpc is enough.  
Maybe later on down the line we would need to expose the actual server instance to allow custom messages though.

Reply:
Ah, I see; I was assuming it'd only allow method calls to methods annotated with JsonRpcMethod, but I guess not. Do we need to support inheriting or could we have InProcLanguageServer implement an interface of our choosing and just expose that?

Reply:
My specific worry being if we have to change the signature for any of these (consider whenever we have to tinker/hack around serialization) that any of that is implicitly a breaking change.

Reply:
Updated, now the factory returns an ILanguageServerTarget instance that implements IAsyncDisposable.